fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/sorting/MemorySortCursor.java

```diff
@@ -73,15 +75,21 @@ public CompletableFuture<RecordCursorResult<V>> onNext() {
         if (iterator != null) {
             return CompletableFuture.completedFuture(nextFromIterator());
         }
-        return sorter.load(inputCursor, minimumKey).thenApply(loadResult -> {
+        return scratchpad.load(inputCursor, minimumKey).thenApply(loadResult -> {
             inputContinuation = loadResult.getSourceContinuation();
             if (loadResult.getSourceNoNextReason().isOutOfBand()) {
                 // The input cursor did not complete; we must save the sorter state in the continuation so can pick up after.
-                MemorySortCursorContinuation<K, V> continuation = new MemorySortCursorContinuation<>(adapter, false, sorter.getMap().values(), minimumKey, inputContinuation);
+                final MemorySortCursorContinuation<K, V> continuation =
+                        new MemorySortCursorContinuation<>(adapter, false, scratchpad.getMap().values(),
+                                adapter.isInsertionOrder() && loadResult.hasSeenMinimumKey()
+                                ? null
+                                : minimumKey,
```

Comment:
I'm not sure the whole "minimum key" thing makes all that much sense for the dam case, though perhaps continuations for dams like this are more nuanced, given that they exist specifically for cases where data has to be loaded into memory, but continuations are designed for incremental processing. In any case, this is another area where having a bit more separation between the in-memory sort and dam code might be nice

Reply:
The minimum key is more like the last consumed item in the dam case.

Reply:
This is now implemented in a different way where all the stuff that is really different between the two cases is extracted into their own `MemorySortComparator`s which are created by the adapter and which can be stateful.